model/network.py

```diff
@@ -407,6 +407,115 @@ def _reset_parameters(self):
             if p.dim() > 1:
                 xavier_uniform_(p)
 
+class GRUSVS_gs(nn.Module):
+    """
+    GRU singing voice synthesis model by Guo Shuai (RUC)
+    """
+
+    def __init__(self, embed_size=512, d_model=512, d_output=1324,
+                 num_layers=2, phone_size=87, n_mels=-1,
+                 dropout=0.1, device="cuda", use_asr_post=False):
+        super(GRUSVS_gs, self).__init__()
+        
+        # Encoder
+        self.embedding_phone = nn.Embedding(phone_size, embed_size)
+        self.rnnEncoder = nn.GRU(embed_size + 2, d_model, bidirectional = True)
+        self.fcEncoder = nn.Linear(d_model * 2, d_model)
+        self.dropoutEncoder = nn.Dropout(dropout)
+
+        # Attention
+        self.attn = nn.Linear((d_model * 2) + d_model, d_model)
+        self.v = nn.Linear(d_model, 1, bias = False)
+
+        # Decoder
+        self.rnnDecoder = nn.GRU((d_model * 2) + d_model, d_model)
+        self.fc_hid1 = nn.Linear((d_model * 2) + d_model + embed_size, 2048)
+        # self.fc_hid2 = nn.Linear(2048, 1600)
+        self.fc_out = nn.Linear(2048, d_output)
+        self.dropoutDecoder = nn.Dropout(dropout)
+        
+
+        self._reset_parameters()
+
+        self.use_asr_post = use_asr_post
+        self.d_model = d_model
+
+    def forward(self, phone, pitch, beats, length_mask):
+        
+        # phone, pitch, beats = [batch size, len, 1]
+        # target = [batch size, max_len, feat_dim]
+
+        batch_size = np.shape(phone)[0]
+        length = torch.max(length_mask,dim=1)[0]    # length = [batch size]
+        
+        phone = phone.permute(1,0,2)
+        pitch = pitch.permute(1,0,2)
+        beats = beats.permute(1,0,2)
+
+        # Encoder 
+        embedded_phone = self.dropoutEncoder(self.embedding_phone(phone[:,:,0])) # # [src len, batch size, emb dim]
+        embedded = torch.cat((embedded_phone.long(), beats.long(), pitch.long()), dim = 2)   # [src len, batch size, emb_dim+2]
+        packed_embedded = nn.utils.rnn.pack_padded_sequence(embedded, length)
+        packed_outputs, hidden = self.rnnEncoder(packed_embedded)
+        encoder_outputs, _ = nn.utils.rnn.pad_packed_sequence(packed_outputs).permute(1, 0, 2) #outputs = [batch size, src len, enc hid dim * 2]
+        hidden = torch.tanh(self.fcEncoder(torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim = 1)))  #hidden = [batch size, dec hid dim]
+
+        # Attention
+        energy = torch.tanh(self.attn(torch.cat((hidden, encoder_outputs), dim = 2))) #energy = [batch size, src len, dec hid dim]
+        attention = self.v(energy).squeeze(2)   #attention = [batch size, src len]
+        attention_weights = F.softmax(attention, dim = 1).unsqueeze(1)  #a = [batch size, 1, src len]
+        
+        # Decoder
+        outputs = torch.zeros(trg_len, batch_size, d_output).to(self.device)
+        input = outputs[0]
+        for t in range(0, trg_len):
```

Comment:
do not use frame-by-frame generation in decoder. It would be very slow in training.

Please directly use teacher-forcing

Reply:
This part can be used when inference